src/Authentication/JWTAuthenticator.php

```diff
@@ -25,201 +34,252 @@
-            ->setIssuedAt(time())
+            ->setIssuedAt($now)
             // Configures the time that the token can be used (nbf claim)
-            ->setNotBefore(time() + $config->get('nbf_time'))
+            ->setNotBefore($now + $config->get('nbf_time'))
             // Configures the expiration time of the token (nbf claim)
-            ->setExpiration(time() + $config->get('nbf_expiration'))
-            // Configures a new claim, called "uid"
-            ->set('uid', $member->ID)
+            ->setExpiration($now + $config->get('nbf_expiration'))
+            // Set renew expiration
+            ->set('rexp', $now + $config->get('nbf_refresh_expiration'))
+            // Configures a new claim, called "rid"
+            ->set('rid', $record->ID)
             // Set the subject, which is the member
             ->setSubject($member->getJWTData())
             // Sign the key with the Signer's key
-            ->sign($this->signer, $this->privateKey);
-
-        // Save the member if it's not anonymous
-        if ($member->ID > 0) {
-            $member->JWTUniqueID = $uniqueID;
-            $member->write();
-        }
+            ->sign($this->getSigner(), $this->getPrivateKey());
 
         // Return the token
         return $token->getToken();
     }
 
     /**
-     * @param string $token
+     * @param string      $token
      * @param HTTPRequest $request
-     * @param ValidationResult $result
-     * @return null|Member
+     * @return array Array with JWTRecord and int status (STATUS_*)
      * @throws BadMethodCallException
      */
-    private function validateToken($token, $request, &$result)
+    public function validateToken(string $token, HTTPrequest $request): array
     {
-        $this->setKeys();
+        // Ensure token given at all
+        if (!$token) {
+            return [null, TokenStatusEnum::STATUS_INVALID];
+        }
+
+        // Parse token
         $parser = new Parser();
-        $parsedToken = $parser->parse((string)$token);
+        try {
+            $parsedToken = $parser->parse($token);
+        } catch (Exception $ex) {
+            // Un-parsable tokens are invalid
+            return [null, TokenStatusEnum::STATUS_INVALID];
+        }
 
-        // Get a validator and the Member for this token
-        list($validator, $member) = $this->getValidator($request, $parsedToken);
+        // Validate token against Id and user-agent
+        $userAgent = $request->getHeader('User-Agent');
+        /** @var JWTRecord $record */
+        $record = JWTRecord::get()
+            ->filter(['UserAgent' => $userAgent])
+            ->byID($parsedToken->getClaim('rid'));
+        if (!$record) {
+            return [null, TokenStatusEnum::STATUS_INVALID];
+        }
 
-        $verified = $parsedToken->verify($this->signer, $this->publicKey);
+        // Get validator for this token
+        $now = DBDatetime::now()->getTimestamp();
+        $validator = new ValidationData();
```

Comment:
`ValidationData::create()`?

Reply:
It's  not a silverstripe class. Can't use that.

Reply:
Okay :)